src/k8s/pkg/k8sd/controllers/feature.go

```diff
@@ -69,65 +69,96 @@ func NewFeatureController(opts FeatureControllerOpts) *FeatureController {
-			return fmt.Errorf("failed to apply DNS configuration: %w", err)
+	go c.reconcileLoop(ctx, getClusterConfig, "dns", c.triggerDNSCh, c.reconciledDNSCh, func(cfg types.ClusterConfig) (types.FeatureStatus, error) {
+		featureStatus, dnsIP, err := features.Implementation.ApplyDNS(ctx, c.snap, cfg.DNS, cfg.Kubelet, cfg.Annotations)
+
+		if err != nil {
+			return featureStatus, fmt.Errorf("failed to apply DNS configuration: %w", err)
 		} else if dnsIP != "" {
 			if err := notifyDNSChangedIP(ctx, dnsIP); err != nil {
-				return fmt.Errorf("failed to update DNS IP address to %s: %w", dnsIP, err)
+				// we already have featureStatus.Message which contains wrapped error of the Apply<Feature>
+				// (or empty if no error occurs). we further wrap the error to add the DNS IP change error to the message
+				changeErr := fmt.Errorf("failed to update DNS IP address to %s: %w", dnsIP, err)
+				featureStatus.Message = fmt.Sprintf("%s: %v", featureStatus.Message, changeErr)
+				return featureStatus, changeErr
 			}
 		}
-		return nil
-	})
+		return featureStatus, nil
+	}, setFeatureStatus)
 }
 
-func (c *FeatureController) reconcile(ctx context.Context, getClusterConfig func(context.Context) (types.ClusterConfig, error), apply func(cfg types.ClusterConfig) error) error {
+func (c *FeatureController) reconcile(
+	ctx context.Context,
+	getClusterConfig func(context.Context) (types.ClusterConfig, error),
+	apply func(cfg types.ClusterConfig) (types.FeatureStatus, error),
+	componentName string,
+	setFeatureStatus func(context.Context, string, types.FeatureStatus) error,
+) error {
 	cfg, err := getClusterConfig(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to retrieve cluster configuration: %w", err)
 	}
 
-	if err := apply(cfg); err != nil {
+	featureStatus, err := apply(cfg)
+	if err != nil {
 		return fmt.Errorf("failed to apply configuration: %w", err)
 	}
+
+	if err := setFeatureStatus(ctx, componentName, featureStatus); err != nil {
+		return fmt.Errorf("failed to set feature status for '%s': %w", err)
+	}
+
 	return nil
 }
 
-func (c *FeatureController) reconcileLoop(ctx context.Context, getClusterConfig func(context.Context) (types.ClusterConfig, error), componentName string, triggerCh chan struct{}, reconciledCh chan<- struct{}, apply func(cfg types.ClusterConfig) error) {
+func (c *FeatureController) reconcileLoop(
+	ctx context.Context,
+	getClusterConfig func(context.Context) (types.ClusterConfig, error),
+	componentName string,
+	triggerCh chan struct{},
+	reconciledCh chan<- struct{},
+	apply func(cfg types.ClusterConfig) (types.FeatureStatus, error),
+	setFeatureStatus func(ctx context.Context, name string, status types.FeatureStatus) error,
+) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-triggerCh:
-			if err := c.reconcile(ctx, getClusterConfig, apply); err != nil {
+			if err := c.reconcile(ctx, getClusterConfig, apply, componentName, setFeatureStatus); err != nil {
```

Comment:
as a generic note: each `reconcile` loop here does not _need_ access to the full `setFeatureStatus`, only a partial with the specified componentName.

instead of passing both `componentName` and `setFeatureStatus`, consider doing something like

```
if err := c.reconcile(ctx, getClusterConfig, apply, func(ctx context.Context, status types.FeatureStatus) error {
  return setFeatureStatus(ctx, componentName, status)
}); err != nil {
  // ...
}
```

So `reconcile` only sees a ["partial"](https://stackoverflow.com/questions/58339040/what-exactly-is-meant-by-partial-function-in-functional-programming), and does not need to concern itself with passing through the componentName to the setFeatureStatus call

Reply:
right, amazing comment, so we're closing over the `componentName` here (forming a closure), right?